metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py

```diff
@@ -256,62 +230,88 @@ def _gen_operation_aspect_workunits(
+        self,
+        query: str,
+        connection: redshift_connector.Connection,
+        all_tables: Dict[str, Dict[str, List[Union[RedshiftView, RedshiftTable]]]],
     ) -> Iterable[RedshiftAccessEvent]:
-        results = engine.execute(query)
-        for row in results:
-            if not self._should_process_row(row):
-                continue
-            if hasattr(row, "_asdict"):
-                # Compatibility with sqlalchemy 1.4.x.
-                row = row._asdict()
-            access_event = RedshiftAccessEvent(**dict(row.items()))
-            # Replace database name with the alias name if one is provided in the config.
-            if self.config.database_alias:
-                access_event.database = self.config.database_alias
-            yield access_event
+        access_events = []
+        cursor = connection.cursor()
+        cursor.execute(query)
+        results = cursor.fetchmany()
+        field_names = [i[0] for i in cursor.description]
+        while results:
+            for row in results:
+                try:
+                    access_event = RedshiftAccessEvent(
+                        userid=row[field_names.index("userid")],
+                        username=row[field_names.index("username")],
+                        query=row[field_names.index("query")],
+                        querytxt=row[field_names.index("querytxt")].strip()
+                        if row[field_names.index("querytxt")]
+                        else None,
+                        tbl=row[field_names.index("tbl")],
+                        database=row[field_names.index("database")],
+                        schema=row[field_names.index("schema")],
+                        table=row[field_names.index("table")],
+                        starttime=row[field_names.index("starttime")],
+                        endtime=row[field_names.index("endtime")],
+                        operation_type=row[field_names.index("operation_type")]
+                        if "operation_type" in field_names
+                        else None,
+                    )
+                except pydantic.error_wrappers.ValidationError as e:
+                    logging.debug(
+                        f"Validation error on access event creation from row {row}. The error was: {e} Skipping ...."
+                    )
+                    self.report.num_usage_stat_skipped += 1
+                    continue
+
+                # Replace database name with the alias name if one is provided in the config.
+                if self.config.database_alias:
+                    access_event.database = self.config.database_alias
+
+                if not self._should_process_event(access_event, all_tables=all_tables):
+                    # TODO: Add to report number of skipped event
+                    self.report.num_usage_stat_skipped += 1
+                    continue
+
+                # yield access_event
+                access_events.append(access_event)
```

Comment:
Is this purposefully not yielding to fetch from the db all at once? The comment and method name make it sound like it should be a generator